platforms/software/dependency-management/src/main/java/org/gradle/internal/component/AbstractNamedConfigurationException.java

```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.component;
+
+/**
+ * Abstract base class for all exceptions thrown during graph resolution
+ * when resolving a dependency specified using an explicitly named configuration.
+ */
+public abstract class AbstractNamedConfigurationException extends AbstractGraphVariantSelectionException {
```

Comment:
❓ What is the point of these abstract base classes for these exceptions? I have learned to be skeptical of abstract classes and (non-interface) inheritance in general, and from what I can tell this abstract class (and the other ones in this hierarchy) are not used practically for encapsulating logic or expressing some interface. To me, these just seem like added complexity without any extra value. 

I could see the advantage that these somehow logically separate the exceptions into groups, where some are graph exceptions and some are artifact exceptions, but does that do anything for us other than serve as documentation? Are we ever making decisions on how to handle an exception based on whether it belongs to one group or another? 

I don't think this is necessarily _wrong_, but I think we should seriously consider whether we are gaining any value from this type hierarchy. Extra code is extra code, and if it does not do anything for us, I'd prefer the route that has less code to maintain

Reply:
I think there are too many exception types in general here and agree there is little need for such a complex hierarchy.

I want to explore simplifying this later, after we have a final(er) Problems API to work with and the WOT is addressed.  I think it would be much better to use 1 or 2 exception classes, with some sort of `FailureType` enum field categorizing the cause.  Right now, the hierarchy helps by relating related causes, making sure all types can supply Suggestions, and (soon) that they all can carry similar failure diagnostic information.

I suppose we could get these benefits without some of the intermediate `abstract` exceptions, but as long as they are empty classes, I don't think there's too much harm in the extra layers for now.

> Are we ever making decisions on how to handle an exception based on whether it belongs to one group or another?

Not currently, but this might be part of simplifying the failure message.  Still exploring.

Reply:
I think this is fine as-is. If you have a vision for this hierarchy then 👍

In general I think we should try to avoid inheritance hierarchies and use more functional approaches for sharing behavior between code, but there is nothing _wrong_ here that I think needs to be changed